### PR TITLE
fix infinite re-renders in portaled `ThemeProvider`

### LINKED ISSUE #2082: Maximum update depth exceeded when `ThemeProvider` is portalled
### Describe the bug (current behavior)

Using a `ReactDOM.createPortal()` with `ThemeProvider` component results in an `Error: Maximum update depth exceeded.` error.
This is initially caught by our e2e tests while trying to bump a minor version of iTwinUI https://github.com/iTwin/appui/pull/855

I'm not certain if the provided minimal repro sample is exactly what we hit in the AppUI repo as it is very tricky to determine what is going on (i.e. removing one of the parent `ThemeProviders` avoids this issue). 

Additionally, memoizing the `ref` of a `Box` used in an iTwinUI `Root` component seems to solve this issue.

### Expected Behavior

`ThemeProvider` should not produce any run-time issues.

### Link to minimal repro

https://stackblitz.com/edit/github-vk35v4?file=src%2FApp.tsx,src%2Fmain.tsx

### Steps To Reproduce

1. Open minimal repro
2. Open dev-tools console
3. Observe the error `Error: Maximum update depth exceeded.`

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by mayank99**
## Changes

This fixes #2082 using a combination of three changes:
1. Memoized the callback ref in `Root` and added an `if` condition. This prevents the infinite loop caused by repeatedly setting state.
2. Moved some set state calls (in `PortalContainer` and `useScopedAtom`) from render into `useEffect`. This avoids some [warnings](https://redirect.github.com/facebook/react/issues/18178) and incorrect states caused by 1, at the cost of requiring an extra render.
3. Added a check for `ownerDocument` before creating portal containers. This is necessary for popout window components, where `ownerDocument` might not be set on initial page load.

**Note**: I started refactoring the state management in `ThemeProvider`, which might avoid these issues altogether, but I'll save it for another PR.

## Testing

Recreated the linked issue in vite playground. Not seeing any more runtime errors and the UI renders correctly. Added basic e2e test to verify this scenario.

## Docs

N/A. Added changeset.

### CODE REVIEW

**Review by Ben-Pusey-Bentley — COMMENTED** [review]
I am still getting the maximum update depth exceeded issue on this branch when using the code from the original sandbox from the linked issue. The e2e test does pass for me, though, and passing the set function to the ref instead of using the `useEffect` also works.
image

**Review by mayank99 — COMMENTED** [review]
@Ben-Pusey-Bentley Good catch! Not sure why the e2e test was still passing, but it should be fixed now after the latest push. I re-verified manually in vite playground.

<details><summary>App.tsx</summary>

```tsx
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Button, ThemeProvider } from '@itwin/itwinui-react';

export default function App() {
  const [el, setEl] = React.useState<HTMLElement | null>(null);

  React.useEffect(() => {
    setEl(document.getElementById('x'));
  }, []);

  return (
    <ThemeProvider>
      <div id='x' />
      {el &&
        ReactDOM.createPortal(
          <ThemeProvider>
            <Button>X</Button>
          </ThemeProvider>,
          el,
        )}
    </ThemeProvider>
  );
}
```

</details>

**Review by Ben-Pusey-Bentley — APPROVED** [review]
Yeah, it works great now. 
image
LGTM!

File: packages/itwinui-react/src/core/ThemeProvider/ThemeProvider.tsx [hunk 1]
```diff
@@ -342,18 +351,25 @@ const PortalContainer = React.memo(
...
 
-    // Synchronize atom with the correct portal container if necessary.
-    if (portalTarget && portalTarget !== portalContainer) {
-      setPortalContainer(portalTarget);
+  // Synchronize atom with the correct portal target if necessary.
+  React.useEffect(() => {
+    if (target && target !== portalContainer) {
+      setPortalContainer(target);
     }
+  });
```
[the remark this answers is not in the document]
Comment (mayank99) [reply packages/itwinui-react/src/core/ThemeProvider/ThemeProvider.tsx:369]: I intentionally wrote it this way. Having a deps array gives the false impression that it's meant to run when the deps change, when in fact we don't really care about any of that.

The important thing to understand is that only the `if` check is running on each render; the code inside does not execute on each render. The only difference from before is that it's now moved out of the rendering path into a `useEffect`.

**Pull request merged**